Added a harvester for the BHL

Eleonore9 opened the pull request:
Added an harvester for the Biodiversity Heritage Library (http://biodivlib.wikispaces.com/Developer+Tools+and+API#Developer%20Tools-OAI-PMH) that uses OAI format.

The names weren't properly parsed so I added functions to customise the parsing.

erinspace commented:
Thanks so much for this PR! Would you mind adding a favicon for cogprints into the https://github.com/fabianvf/scrapi/tree/develop/img/favicons folder

fabianvf commented on scrapi/processing/storage.py:
```diff
@@ -12,6 +12,7 @@
 from scrapi.processing.base import BaseProcessor
 
 
+
```
PEP-E303 too many blank lines (3). 
Python style guides recommend 2 newlines between imports and the rest of the logic. We do perform automated code quality checks with the tool ```flake8```. Unfortunately, this does mean that your build will fail in cases where the style guide isn't followed. I'll comment throughout the rest of the repository where style errors are found, but to get the list of errors you can also run ``` flake8 .``` from the scrapi project to get the errors locally. If you would rather not fix these errors, let me know and I can pull in your branch and make the changes myself.

erinspace commented:
Hah, we originally did this to add something innocuous to let us successfully rebase, as there were originally just 2 lines. I'd also be happy to make this change, along with the favicon, so let me know!

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+"""Harvests Biodiversity Heritage Library OAI Repository (BHL) metadata for ingestion 
```
W291 trailing whitespace
Python style guidelines recommend that you trim trailing whitespace at the end of lines

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+                names.append(name)
+        elif arg:
+            names.append(arg)
+            
```
W293 blank line contains whitespace
Blank lines should have no whitespace in them

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+        elif arg:
+            names.append(arg)
+            
+    # Filter people names and clean dates and extra spaces        
```
W291 trailing whitespace
Python style guidelines recommend that you trim trailing whitespace at the end of lines

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+    people = [re.sub(r'\d+-(\d+)?', r'', n).strip() for n in filter(lambda x: ', ' in x, names)] or []
+    # Filter institution names and clean tabs
+    inst = [re.sub(r'\\t', r'', n).strip() for n in filter(lambda x: ', ' not in x, names) or []]
+        
```
W293 blank line contains whitespace
Blank lines should have no whitespace in them

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+    # Filter institution names and clean tabs
+    inst = [re.sub(r'\\t', r'', n).strip() for n in filter(lambda x: ', ' not in x, names) or []]
+        
+    if len(people)  > 0 and len(inst) > 0:
```
E221 multiple spaces before operator
The first ```>``` operator has two spaces in front of it, python style guidelines recommend 1 space before and after each operator.

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+        return default_name_parser(people) + institution_name_parser(inst)
+    elif len(inst) > 0:
+        return institution_name_parser(inst)
+    if len(people)  > 0:
```
E221 multiple spaces before operator
The first > operator has two spaces in front of it, python style guidelines recommend 1 space before and after each operator.

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+    else:
+        return ''
+
+    
```
W293 blank line contains whitespace
Blank lines should have no whitespace in them

fabianvf commented on scrapi/harvesters/bhl.py:
```diff
@@ -0,0 +1,64 @@
+    schema = updated_schema(OAISCHEMA, {
+        'contributors': ('//dc:creator/node()', '//dc:contributor/node()', aoi_process_contributors_bhl)
+    })
+
```
W391 blank line at end of file
According to python style guidelines, the extra newlines at the end should be removed

fabianvf commented:
Just finished the code review, other than the issues I pointed out, it looks great! Thanks for hacking with us. Again, if you would rather not deal with the style stuff, just let me know and I can make these modifications.

Eleonore9 commented:
Hi! Thanks for the comments @fabianvf ! I'd like to fix the style issues.
 I just underestimated the effect of jetlag, but I'll get back to it tomorrow =)

Eleonore9 commented:
Fixed flake8 issues and added favicon! The build is in progress...

Thanks for being at PyCon and for organising a sprint! I had come across OSF on the internet and I'm super interested by open science (as a programmer and former biologist). I'll do my best to contribute more =)

erinspace commented:
@Eleonore9 this is awesome!! Thanks so much again for sprinting with us, was nice talking to you and working through some of the sticking points in the harvester creation process! Looking forward to working together more in the future!
